src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,357 @@
+        {
+            ReturnErrorOnFailure(
+                out.Append({ path.mClusterId, serverCluster->GetDataVersion(path), serverCluster->GetClusterFlags(path) }));
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::GeneratedCommands(const ConcreteClusterPath & path, ReadOnlyBufferBuilder<CommandId> & out)
+{
+    ServerClusterInterface * serverCluster = GetServerCluster(path.mEndpointId, path.mClusterId, nullptr);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->GeneratedCommands(path, out);
+}
+CHIP_ERROR CodeDrivenDataModelProvider::AcceptedCommands(const ConcreteClusterPath & path,
+                                                         ReadOnlyBufferBuilder<DataModel::AcceptedCommandEntry> & out)
+{
+    ServerClusterInterface * serverCluster = GetServerCluster(path.mEndpointId, path.mClusterId, nullptr);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->AcceptedCommands(path, out);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Attributes(const ConcreteClusterPath & path,
+                                                   ReadOnlyBufferBuilder<DataModel::AttributeEntry> & out)
+{
+    ServerClusterInterface * serverCluster = GetServerCluster(path.mEndpointId, path.mClusterId, nullptr);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->Attributes(path, out);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::EventInfo(const ConcreteEventPath & path, DataModel::EventEntry & eventInfo)
+{
+    ServerClusterInterface * serverCluster = GetServerCluster(path.mEndpointId, path.mClusterId, nullptr);
+    VerifyOrReturnError(serverCluster != nullptr, CHIP_ERROR_NOT_FOUND);
+    return serverCluster->EventInfo(path, eventInfo);
+}
+
+void CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged(const AttributePathParams & path)
+{
+    if (mContext.dataModelChangeListener == nullptr)
+    {
+        ChipLogError(DataManagement, "Temporary_ReportAttributeChanged called before provider has been started.");
+        return;
+    }
+    mContext.dataModelChangeListener->MarkDirty(path);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::AddEndpoint(EndpointInterfaceRegistration & registration)
+{
+    VerifyOrReturnError(registration.endpointInterface != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    ReturnErrorOnFailure(mEndpointInterfaceRegistry.Register(registration));
+
+    // If the provider has not been started, return early, no need to startup clusters.
+    VerifyOrReturnError(mContext.dataModelChangeListener != nullptr, CHIP_NO_ERROR);
+
+    // If the provider has already been started (its mContext is initialized) AND its server
+    // cluster context is also set up, then the clusters for this new endpoint should be started.
+    VerifyOrReturnError(mServerClusterContext.has_value(), CHIP_ERROR_INCORRECT_STATE);
```

Comment:
This does not match the conditions in shutdown, right?  That does not check any of these things.

Nor does it match the conditions in RemoveEndpoint, which checks only mContext.dataModelChangeListener.

It also does not match the behavior of CodeDrivenDataModelProvider::Startup, which does not check mContext.dataModelChangeListener is not null.

All these checks have to be consistent with each other; we need a single source of truth for "we are in a state where we should be trying to start up things".